packages/react-native-audio-api/android/src/main/cpp/audioapi/android/core/utils/ffmpegBackend/FFmpegFileWriter.cpp

```diff
@@ -0,0 +1,468 @@
+  }
+
+  AVChannelLayout inChannelLayout;
+  av_channel_layout_default(&inChannelLayout, inputChannels);
+
+  av_opt_set_chlayout(resampleCtx_.get(), "in_chlayout", &inChannelLayout, 0);
+  av_opt_set_chlayout(
+      resampleCtx_.get(), "out_chlayout", &encoderCtx_->ch_layout, 0);
+
+  av_opt_set_int(resampleCtx_.get(), "in_sample_rate", inputRate, 0);
+  av_opt_set_int(
+      resampleCtx_.get(), "out_sample_rate", encoderCtx_->sample_rate, 0);
+
+  av_opt_set_sample_fmt(
+      resampleCtx_.get(), "in_sample_fmt", AV_SAMPLE_FMT_FLT, 0);
+  av_opt_set_sample_fmt(
+      resampleCtx_.get(), "out_sample_fmt", encoderCtx_->sample_fmt, 0);
+
+  int result = swr_init(resampleCtx_.get());
+
+  if (result < 0) {
+    return ReturnStatus<void>::Error(
+        "Failed to initialize resampler for file: " + parseErrorCode(result));
+  }
+
+  return ReturnStatus<void>::Success();
+}
+
+void FFmpegAudioFileWriter::initializeBuffers(int32_t maxBufferSize) {
+  frame_ = av_unique_ptr<AVFrame>(av_frame_alloc());
+  packet_ = av_unique_ptr<AVPacket>(av_packet_alloc());
+
+  int contextFrameRatio = 4;
+
+  if (encoderCtx_->frame_size > 0) {
+    contextFrameRatio = static_cast<int>(std::ceil(
+        static_cast<double>(maxBufferSize) /
+        static_cast<double>(encoderCtx_->frame_size)));
+  }
+
+  int calculatedSize =
+      (encoderCtx_->frame_size > 0 ? encoderCtx_->frame_size * contextFrameRatio
+                                   : maxBufferSize * contextFrameRatio);
+
+  int fifoSize = std::max(calculatedSize, 8192);
+
+  audioFifo_ = av_unique_ptr<AVAudioFifo>(av_audio_fifo_alloc(
+      encoderCtx_->sample_fmt, encoderCtx_->ch_layout.nb_channels, fifoSize));
+}
+
+bool FFmpegAudioFileWriter::resampleAndPushToFifo(
+    void *inputData,
+    int inputFrameCount) {
+  int result = 0;
+  int outputLength = av_rescale_rnd(
+      inputFrameCount,
+      encoderCtx_->sample_rate,
+      streamSampleRate_,
+      AV_ROUND_UP);
+
+  result = prepareFrameForEncoding(outputLength);
+
+  if (result < 0) {
+    return false;
+  }
+
+  const uint8_t *inputs[1] = {reinterpret_cast<const uint8_t *>(inputData)};
+
+  int convertedSamples = swr_convert(
+      resampleCtx_.get(), frame_->data, outputLength, inputs, inputFrameCount);
+
+  if (convertedSamples < 0) {
+    av_frame_unref(frame_.get());
+    return false;
+  }
+
+  int written = av_audio_fifo_write(
+      audioFifo_.get(), (void **)frame_->data, convertedSamples);
+
+  if (written < convertedSamples) {
+    return false;
+  }
+
+  return true;
+}
+
+int FFmpegAudioFileWriter::processFifo(bool flush) {
+  int result = 0;
+  int frameSize = encoderCtx_->frame_size > 0 ? encoderCtx_->frame_size : 512;
+
+  while (av_audio_fifo_size(audioFifo_.get()) >= (flush ? 1 : frameSize)) {
+    const int chunkSize =
+        std::min(av_audio_fifo_size(audioFifo_.get()), frameSize);
+
+    if (prepareFrameForEncoding(chunkSize) < 0) {
+      return -1;
```

Comment:
it might be better to return some AV_ERROR code (f.e. unknown) for the parser used higher up in the call stack